content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+   Custom start=0x00000078 end=0x0005e02e (size=0x0005dfb6) ".debug_info"
+   Custom start=0x0005e031 end=0x0005e197 (size=0x00000166) ".debug_pubtypes"
+   Custom start=0x0005e19b end=0x00087051 (size=0x00028eb6) ".debug_ranges"
+   Custom start=0x00087054 end=0x00087fef (size=0x00000f9b) ".debug_abbrev"
+   Custom start=0x00087ff3 end=0x000cf974 (size=0x00047981) ".debug_line"
+   Custom start=0x000cf978 end=0x00167aa8 (size=0x00098130) ".debug_str"
+   Custom start=0x00167aac end=0x0019f276 (size=0x000377ca) ".debug_pubnames"
+   Custom start=0x0019f278 end=0x0019f299 (size=0x00000021) "name"
+   Custom start=0x0019f29b end=0x0019f2e8 (size=0x0000004d) "producers"
+```
+
+`wasm-objdump` is quite versatile and offers a familiar CLI for people who have experience developing for other ISAs in assembly. However, specifically for debugging binary size, it lacks simple helpers like odering the sections by size. Luckily, there is another WebAssembly-specific tool called [Twiggy], that excels at this:
+
+```
+$ twiggy top target/wasm32-unknown-unknown/release/my_project.wasm
+ Shallow Bytes │ Shallow % │ Item
+───────────────┼───────────┼─────────────────────────────────────────
+        622885 ┊    36.63% ┊ custom section '.debug_str'
+        384938 ┊    22.64% ┊ custom section '.debug_info'
+        293237 ┊    17.24% ┊ custom section '.debug_line'
+        227258 ┊    13.36% ┊ custom section '.debug_pubnames'
+        167592 ┊     9.85% ┊ custom section '.debug_ranges'
+          3981 ┊     0.23% ┊ custom section '.debug_abbrev'
+           342 ┊     0.02% ┊ custom section '.debug_pubtypes'
+            67 ┊     0.00% ┊ custom section 'producers'
+            25 ┊     0.00% ┊ custom section 'name' headers
+            20 ┊     0.00% ┊ custom section '.debug_pubnames' headers
+            19 ┊     0.00% ┊ custom section '.debug_pubtypes' headers
+            18 ┊     0.00% ┊ custom section '.debug_ranges' headers
+            17 ┊     0.00% ┊ custom section '.debug_abbrev' headers
+            16 ┊     0.00% ┊ custom section '.debug_info' headers
+            16 ┊     0.00% ┊ custom section '.debug_line' headers
+            15 ┊     0.00% ┊ custom section '.debug_str' headers
+            14 ┊     0.00% ┊ export "__heap_base"
+            13 ┊     0.00% ┊ export "__data_end"
+            12 ┊     0.00% ┊ custom section 'producers' headers
+             9 ┊     0.00% ┊ export "memory"
+             9 ┊     0.00% ┊ add
+...
+```
+
+It’s now clearly visible that all main contributors to the module size are custom sections, which — by definition — are not relevant for the execution of the module. Their names imply that they contain information that is used for debugging, so it is somewhat surprising they are even emitted for a `--release` build. I suspect that release builds make `rustc` turn on code optimization, remove asserts and other checks. It does not affect whether debug symbols are packed into the binary.
```

Comment:
So what happens here is, `--release` actually does disable debug symbols for compilation, but the stdlib is already precompiled with them included. When they're linked together, you get the bloat of the stdlib debug symbols, but none of your own app which is uh the worst of both worlds. It's an old bug.

What's worth mentioning here is a `strip` Cargo feature, which will do the debug stripping for you regardless of the compilation target:

```rust
[profile.release]
strip = true
```

https://doc.rust-lang.org/cargo/reference/profiles.html#strip

Reply:
Good to know tho. Although you also lose function names.

Related, which you may (or may not) find interesting: I opened this PR to `wasm-strip` so you can keep individual custom sections around: https://github.com/WebAssembly/wabt/pull/2143